src/napari_matplotlib/histogram.py

```diff
@@ -63,24 +142,83 @@ def _update_contrast_lims(self) -> None:
 
         self.figure.canvas.draw()
 
-    def draw(self) -> None:
-        """
-        Clear the axes and histogram the currently selected layer/slice.
-        """
-        layer = self.layers[0]
+    def autoset_widget_bins(self, data: npt.NDArray[Any]) -> None:
+        """Update widgets with bins determined from the image data"""
+        bins = _get_bins(data)
 
+        # Disable callbacks whilst setting widget values
+        for widget in self._bin_widgets.values():
+            widget.blockSignals(True)
+
+        self.bins_start = bins[0]
+        self.bins_stop = bins[-1]
+        self.bins_num = bins.size - 1
+
+        for widget in self._bin_widgets.values():
+            widget.blockSignals(False)
+
+    @property
+    def bins_start(self) -> float:
+        """Minimum bin edge"""
+        return self._bin_widgets["start"].value()
+
+    @bins_start.setter
+    def bins_start(self, start: Union[int, float]) -> None:
+        """Set the minimum bin edge"""
+        self._bin_widgets["start"].setValue(start)
+
+    @property
+    def bins_stop(self) -> float:
+        """Maximum bin edge"""
+        return self._bin_widgets["stop"].value()
+
+    @bins_stop.setter
+    def bins_stop(self, stop: Union[int, float]) -> None:
+        """Set the maximum bin edge"""
+        self._bin_widgets["stop"].setValue(stop)
+
+    @property
+    def bins_num(self) -> int:
+        """Number of bins to use"""
+        return self._bin_widgets["num"].value()
+
+    @bins_num.setter
+    def bins_num(self, num: int) -> None:
+        """Set the number of bins to use"""
+        self._bin_widgets["num"].setValue(num)
+
+    def _get_layer_data(self, layer: napari.layers.Layer) -> npt.NDArray[Any]:
+        """Get the data associated with a given layer"""
         if layer.data.ndim - layer.rgb == 3:
             # 3D data, can be single channel or RGB
             data = layer.data[self.current_z]
             self.axes.set_title(f"z={self.current_z}")
         else:
             data = layer.data
+
         # Read data into memory if it's a dask array
         data = np.asarray(data)
 
+        return data
+
+    def draw(self) -> None:
+        """
+        Clear the axes and histogram the currently selected layer/slice.
+        """
+        layer = self.layers[0]
+        data = self._get_layer_data(layer)
+
         # Important to calculate bins after slicing 3D data, to avoid reading
         # whole cube into memory.
-        bins = _get_bins(data)
+        if data.dtype.kind in {"i", "u"}:
+            # Make sure integer data types have integer sized bins
+            step = abs(self.bins_stop - self.bins_start) // (self.bins_num)
+            step = max(1, step)
+            bins = np.arange(self.bins_start, self.bins_stop + step, step)
```

Comment:
I think it would be good to show a notification when the actual number of bins used differs from that requested in the widgets, but perhaps this could be a different PR?

Reply:
I've added the notification now, but happy to remove it and add separately if it's easier